test/test_cuda.py

```diff
@@ -1901,6 +1901,237 @@ def worker(rank):
 t2.start()
 """])
 
+    def test_grad_scaling_builtins(self):
+        rscale = torch.cuda.FloatTensor([0.25])
+
+        found_inf = torch.cuda.FloatTensor([0.0])
+        g = torch.cuda.FloatTensor(10).fill_(4.0)
+        torch._amp_unscale_inf_check_(g, rscale, found_inf)
+        self.assertTrue(found_inf.item() == 0.0)
+        self.assertTrue(torch.allclose(g, torch.ones(10, device="cuda", dtype=torch.float32), atol=1e-7))
+
+        found_inf.zero_()
+        g = torch.cuda.FloatTensor([float('inf')])
+        torch._amp_unscale_inf_check_(g, rscale, found_inf)
+        self.assertTrue(found_inf.item() == 1.0)
+
+        found_inf.zero_()
+        g = torch.cuda.FloatTensor([float('nan')])
+        torch._amp_unscale_inf_check_(g, rscale, found_inf)
+        self.assertTrue(found_inf.item() == 1.0)
+
+        growth_factor = 4.0
+        backoff_factor = 0.5
+        current_scale = torch.cuda.FloatTensor([4.0])
+
+        found_inf.zero_()
+        new_scale = torch._amp_update_scale(current_scale, found_inf, growth_factor, backoff_factor)
+        self.assertTrue(new_scale.item(), 16.0)
+
+        found_inf.fill_(1.0)
+        new_scale = torch._amp_update_scale(current_scale, found_inf, growth_factor, backoff_factor)
+        self.assertTrue(new_scale.item(), 2.0)
+
+    def _create_scaling_models_optimizers(self):
+        # Create a module+optimizer that will use scaling, and a control module+optimizer
+        # that will not use scaling, against which the scaling-enabled module+optimizer can be compared.
+        mod_control = torch.nn.Sequential(torch.nn.Linear(8, 8), torch.nn.Linear(8, 8)).cuda()
+        mod_scaling = torch.nn.Sequential(torch.nn.Linear(8, 8), torch.nn.Linear(8, 8)).cuda()
+        for c, s in zip(mod_control.parameters(), mod_scaling.parameters()):
+            s.data.copy_(c.data)
+
+        opt_control = torch.optim.SGD(mod_control.parameters(), lr=1.0)
+        opt_scaling = torch.optim.SGD(mod_scaling.parameters(), lr=1.0)
+
+        return mod_control, mod_scaling, opt_control, opt_scaling
+
+    def _create_scaling_case(self):
+        data = [(torch.randn((8, 8), device="cuda"), torch.randn((8, 8), device="cuda")),
+                (torch.randn((8, 8), device="cuda"), torch.randn((8, 8), device="cuda")),
+                (torch.randn((8, 8), device="cuda"), torch.randn((8, 8), device="cuda")),
+                (torch.randn((8, 8), device="cuda"), torch.randn((8, 8), device="cuda"))]
+
+        loss_fn = torch.nn.MSELoss().cuda()
+
+        skip_iter = 3
+
+        return self._create_scaling_models_optimizers() + (data, loss_fn, skip_iter)
+
+    # Genericize a bit of the testing logic to avoid too much copy-pasting below
```

Comment:
Is this a TODO?

Reply:
No, just a note.  I can delete if necessary.